src/couch/src/couch_db.erl

```diff
@@ -397,34 +397,43 @@ when
     PurgeOption :: interactive_edit | replicated_changes,
     Reply :: {ok, []} | {ok, [Rev]}.
 purge_docs(#db{main_pid = Pid} = Db, UUIDsIdsRevs, Options) ->
-    UUIDsIdsRevs2 = [
+    UUIDsIdsRevs1 = [
         {UUID, couch_util:to_binary(Id), Revs}
      || {UUID, Id, Revs} <- UUIDsIdsRevs
     ],
-    % Check here if any UUIDs already exist when
-    % we're not replicating purge infos
-    IsRepl = lists:member(replicated_changes, Options),
-    if
-        IsRepl ->
-            ok;
-        true ->
-            UUIDs = [UUID || {UUID, _, _} <- UUIDsIdsRevs2],
-            lists:foreach(
-                fun(Resp) ->
-                    if
-                        Resp == not_found ->
-                            ok;
-                        true ->
-                            Fmt = "Duplicate purge info UIUD: ~s",
-                            Reason = io_lib:format(Fmt, [element(2, Resp)]),
-                            throw({badreq, Reason})
-                    end
-                end,
-                get_purge_infos(Db, UUIDs)
-            )
+    % Gather any existing purges with the same UUIDs
+    UUIDs = element(1, lists:unzip3(UUIDsIdsRevs1)),
+    Old1 = get_purge_infos(Db, UUIDs),
+    Old2 = maps:from_list([{UUID, {Id, Revs}} || {_, UUID, Id, Revs} <- Old1]),
+    % Filter out all the purges which have already been processed
+    FilterCheckFun = fun({UUID, Id, Revs}) ->
+        case maps:is_key(UUID, Old2) of
+            true ->
+                #{UUID := {OldId, OldRevs}} = Old2,
+                case Id =:= OldId andalso Revs =:= OldRevs of
```

Comment:
Should we usort our revs here?

Reply:
Might not be a bad idea. sort or usort? we don't assert uniqueness I think in other places

Reply:
Used `lists:sort/1`. Maybe we should just lists:usort all the incoming purge requests?

Reply:
Opted to usort both in fabric and in couch_db. But also kept the sort in the duplicate request check, just in case there are already existing purges with duplicate revs inserted before upgrading to this PR.